ISY994 sensor improvements

OverloadUT opened the pull request:
## Description:
### Binary Sensor Changes
Today, using Insteon sensors via the ISY994 platform is a pretty big pain. Issues:
1. The user has to name the sensors very specifically inside ISY994 based on the Hass config option.
2. The user has to specify the device_class manually using `customize:`
3. If the hardware sensor is configured to use unique nodes for on and off (which allows more powerful scene configuration in the Insteon ecosystem) then the sensor entities _will not function whatsoever_ inside Hass, requiring the user to set up a "virtual entity" using the Programs feature in the ISY994 platform. While functional, these virtual entities update slower than native devices and require a lot of user configuration inside the ISY994 software.
4. More recent Insteon hardware report a daily heartbeat for the purpose of determining when a battery is dead, and there is no way to see this inside Hass due to it being an entity that never changes status.

This overhaul eliminates all of those pain points, and make all Insteon ISY994 sensors work out of the box with the following improvements:
1. The multiple ISY994 "nodes" used by a sensor no longer each get unique entities in Hass, but instead all attach to a single `binary_sensor`.
2. The `binary_sensor` logic is now smart enough to handle BOTH ways that hardware can report changes: Either by the primary node changing status, OR by the primary and negative nodes reporting ON events.
3. All Insteon ISY994 sensor classes are now automatically determined by looking at the ISY994 "type" attribute, which indicates the exact hardware in use.
4. The timestamp of the last heartbeat is now stored as a sensor attribute, allowing for automations to trigger on a missed heartbeat. On boot-up, the attribute will be set to `STATE_UNKNOWN` _if and only if_ a heartbeat node is actually present for the device. This means that older sensors without the heartbeat feature will not get erroneously identified as missing heartbeats (if the automation is written correctly - I will provide an example in the updated docs)!

### Event Support
This PR also adds ISY994 Control event support: all Insteon "control" events will now emit as Hass events. This allows automations based on button presses, double-presses, long-holds, and others!

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4207

## Breaking Change Info
ISY994 sensor devices previously showed up as switches by default, and now they will show up as binary_sensors. Any automations set up to trigger based on those old switch versions should be updated to the binary_sensor versions. _Note: For now, this only applies to people using the 5.x beta versions of the ISY firmware._

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

OverloadUT committed:
Fire events for ISY994 control events

This allows hass to react directly to Insteon button presses (on switches and remotes), including presses, double-presses, and long holds

OverloadUT committed:
Overhaul binary sensors in ISY994 to be functional "out of the box"

We now smash all of the subnodes from the ISY994 in to one Hass binary_sensor, and automatically support both paradigms of state reporting that Insteon sensors can do. Sometimes a single node's state represents the sensor's state, other times two nodes are used and only "ON" events are sent from each. The logic between the two forunately do not conflict so we can support both without knowing which mode the device is in.

This also allows us to handle the heartbeat functionality that certain sensors have - we simply store the timestamp of the heartbeat as an attribute on the sensor device. It defaults to Unknown on bootup if and only if the device supports heartbeats, due to the presence of subnode 4.

OverloadUT committed:
Parse the binary sensor device class from the ISY's device "type"

Now we automatically know which sensors are moisture, motion, and openings! (We also reverse the moisture sensor state, because Insteon reports ON for dry on the primary node.)

MartinHjelmare commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +68,154 @@ def setup_platform(hass, config: ConfigType,
     def __init__(self, node) -> None:
         """Initialize the ISY994 binary sensor device."""
         isy.ISYDevice.__init__(self, node)
+        self._child_nodes = []
```
Where is this used?

OverloadUT replied:
Hm, I had intended to store the child nodes in this array, but then I suppose I never actually ended up needing to. It feels weird not storing them, but since it would not be used in the current implementation I'll just remove this!

OverloadUT commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +68,154 @@ def setup_platform(hass, config: ConfigType,
+        elif subnode_id == 4:
+            # Heartbeat node that just reports "On" every 24 hours
+            child.controlEvents.subscribe(self._heartbeat_node_control_handler)
+            self._heartbeat_timestamp = STATE_UNKNOWN
```
It's important for there to be a difference between the attribute not existing (None) and the attribute currently being "Unknown" because we haven't seen a heartbeat since the last boot. If it was simply None until the first heartbeat, an automation can't tell if this device ever SHOULD give a heartbeat.

OverloadUT commented on homeassistant/components/isy994.py:
```diff
@@ -231,11 +241,22 @@ def __init__(self, node) -> None:
             'changed', self.on_update)
 
+        if hasattr(self._node, 'controlEvents'):
+            self._control_handler = self._node.controlEvents.subscribe(
```
You know what, it didn't occur to me that this PR is actually branched off of PR #10664, which is where this code comes from. That being said, I'll make the change to the base PR and rebase this one to that.

OverloadUT replied:
(This is now done even though GitHub is not recognizing this comment as outdated)

OverloadUT commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +68,154 @@ def setup_platform(hass, config: ConfigType,
+        """Get the state attributes for the device."""
+        attr = super(ISYBinarySensorDevice, self).device_state_attributes
+        if self._heartbeat_timestamp is not None:
+            attr['last_heartbeat'] = self._heartbeat_timestamp
```
I don't see the alternative here. Heartbeats are daily pings and that's it. The idea here is that you could write some automatons that sent an alert if the heartbeat timestamp becomes more than 24 hours old.

I'm certainly open to suggestions but this seemed like the most straightforward way to present it to the user. I thought about making it some sort of entity that handled the 24 hour countdown internally, but that would depend on all Insteon devices always being on a 24 hour heartbeat, which is an assumption I don't think we should make...

OverloadUT commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +68,154 @@ def setup_platform(hass, config: ConfigType,
 
     def __init__(self, node) -> None:
         """Initialize the ISY994 binary sensor device."""
         isy.ISYDevice.__init__(self, node)
```
Done

OverloadUT commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +68,154 @@ def setup_platform(hass, config: ConfigType,
+        try:
+            device_type = self._node.type
+        except AttributeError:
+            self._device_class_from_type = None
```
Whoops; artifacts from refactoring this in to its own method.

OverloadUT commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +68,154 @@ def setup_platform(hass, config: ConfigType,
+
+        split_type = device_type.split('.')
+        _LOGGER.debug(split_type)
+        for cls, ids in ISY_DEVICE_TYPES.items():
```
Changed to `device_class`

OverloadUT commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +68,154 @@ def setup_platform(hass, config: ConfigType,
+        try:
+            if self.device_class == 'moisture':
+                return not self._computed_state
+        except AttributeError:
```
If you check the ISY_DEVICE_TYPES const up top, `moisture` is one of the keys

OverloadUT commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +68,154 @@ def setup_platform(hass, config: ConfigType,
+    @property
+    def device_state_attributes(self):
+        """Get the state attributes for the device."""
+        attr = super(ISYBinarySensorDevice, self).device_state_attributes
```
Done

OverloadUT committed:
Code review tweaks

The one material change here is that the event subscribers were moved to the `async_added_to_hass` method, as the handlers depend on things that only exist after the entity has been added.

OverloadUT committed:
Move change event subscription to after entity is added to hass

The event handler method requires `self.hass` to exist, which doesn't have a value until the async_added_to_hass method is called. Should eliminate a race condition.

OverloadUT commented on homeassistant/components/isy994.py:
```diff
@@ -231,11 +241,22 @@ def __init__(self, node) -> None:
         self._change_handler = self._node.status.subscribe(
```
Done! Good catch to eliminate a race condition. This issue was present before as well!

OverloadUT committed:
Handle cases where a sensor's state is unknown

When the ISY first boots up, if a battery-powered sensor has not reported in yet (due to heartbeat or a change in state), the state is unknown until it does.

MartinHjelmare commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +70,183 @@ def setup_platform(hass, config: ConfigType,
+    @asyncio.coroutine
+    def async_added_to_hass(self) -> None:
+        """Subscribe to the node and subnode event emitters."""
+        super().async_added_to_hass()
```
This is a coroutine, so you should `yield from` it to schedule it.

OverloadUT replied:
Thanks for the help here - I am a little unclear on when to yield in a coroutine. Are you saying that I should `yield from` the super() call, but the other calls in this method are fine?

MartinHjelmare replied:
Yes!

MartinHjelmare commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +70,183 @@ def setup_platform(hass, config: ConfigType,
+    def state(self):
+        """Return the state of the binary sensor."""
+        if self._computed_state is None:
+            return STATE_UNKNOWN
```
Return `None`. The base entity class will handle unknown states.

OverloadUT replied:
Hm, I see that the `async_update_ha_state` will indeed do this, but I feel like if we _know_ the state is unknown, this method return `STATE_UNKNOWN` to be explicit. That's what the Entity version of this method does, so the established paradigm is to fall back on `STATE_UNKNOWN`.

MartinHjelmare replied:
No, we should let the entity base class handle this.

MartinHjelmare commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +70,183 @@ def setup_platform(hass, config: ConfigType,
+        try:
+            if self.device_class == 'moisture':
+                return not self._computed_state
+        except AttributeError:
```
What object attribute access could lead to attribute error?

OverloadUT replied:
Fixed - artifact of a refactor

MartinHjelmare reviewed (commented):
I think you should ask @balloob about the timestamp in the state attributes, and possible alternatives.

OverloadUT committed:
Clean up from code review

Fix coroutine await, remove unnecessary exception check, and return None when state is unknown

OverloadUT commented:
To summarize the question that needs @balloob's input:

Some battery-powered Insteon devices send a heartbeat signal every 24 hours as a mechanism to tell if a battery has died. Currently I have this implemented as:
* A heartbeat attribute that gets updated to the current time (`datetime.now().isoformat()`) whenever the heartbeat signal is received.
* If we know the device is heartbeat-capable but haven't seen one yet (on a Hass reboot) then the attribute gets set to STATE_UNKNOWN (allowing an automation to trigger if it stays UNKNOWN for >24 hours, for example)
* If we know the device is not heartbeat-capable, the attribute does not exist in the first place

@MartinHjelmare pointed out that timestamps in attributes is frowned-upon, but I'm not immediately seeing another approach here that keeps the edge case handling. What do you think? Is there any heartbeat precedent in other components?

pvizeli commented:
I set it to WIP. You can remove it after you implement that binary and remove the timestamp from attributes :+1:

OverloadUT committed:
Unknown value from PyISY is now -inf rather than -1

OverloadUT committed:
Move heartbeat handling to a separate sensor

Now all heartbeat-compatible sensors will have a separate `binary_sensor` device that represents the battery state (on = dead)

MartinHjelmare commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +83,273 @@ def setup_platform(hass, config: ConfigType,
+        try:
+            self._negative_node.controlEvents.subscribe(
+                self._negative_node_control_handler)
+        except AttributeError:
```
Why not check `if self._negative_node is not None` instead?

OverloadUT replied:
I went with the Python pattern of "ask for forgiveness" for no other reason than it seems like the Pythonic way. Happy to change it, and I'd love to hear if you have thoughts about when `except` should be used vs checking for None.

MartinHjelmare replied:
If there's a simple check that is safe and always works, I go with an `if`. If there are multiple checks needed or if those checks don't always work, or if cases are unpredictable, I go with `try... except`. But it all depends... :smile:
Here the case looks contained to me since we initialize the attribute to `None`.

OverloadUT replied:
Got it! Since we control that property in a closed environment, we should just check for values that we know it can be. I just pushed the change for these comments

MartinHjelmare commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +83,273 @@ def setup_platform(hass, config: ConfigType,
+        """Send a heartbeat to our heartbeat device, if we have one."""
+        try:
+            self._heartbeat_device.heartbeat()
+        except AttributeError:
```
Why not check `if self._heartbeat_device is not None` instead?

OverloadUT replied:
Same comment as above

MartinHjelmare reviewed (commented):
Overall looks good. Some minor questions/comments below.

OverloadUT committed:
Add support for Unknown state, which is being added in next PyISY 

PyISY will report unknown states as the number "-inf". This is implemented in the base ISY994 component, but subcomponents that override the `state` method needed some extra logic to handle it as well.

OverloadUT committed:
Merge branch 'dev' into isy994-control-events

OverloadUT committed:
Merge branch 'isy994-control-events' into isy994_sensor_improvements

OverloadUT commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -49,22 +83,273 @@ def setup_platform(hass, config: ConfigType,
+    def _detect_device_type(self) -> str:
+        try:
+            device_type = self._node.type
+        except AttributeError:
```
This is to protect against the `self._node.type` not being set. It comes from a third party API that itself can be populated by fourth party code, so just being defensive here!

OverloadUT commented:
I merged in some changes from PR #10664 which this is branched off of

OverloadUT committed:
Change a couple try blocks to explicit None checks

OverloadUT committed:
Bump PyISY to 1.1.0, now that it has been published!

OverloadUT committed:
Merge branch 'isy994-control-events' into isy994_sensor_improvements

OverloadUT committed:
Remove -inf checking from base component

The implementation of the -inf checking was improved in another branch which has been merged in to this branch already.

MartinHjelmare commented on homeassistant/components/isy994.py:
```diff
@@ -270,6 +300,21 @@ def value(self) -> object:
+        if self.is_unknown():
+            return None
+        else:
+            return super().state
```
When would this be used? All isy device classes except binary sensor program seem to override the `state` property anyway. If this call would go to the `Entity` class, it would return `STATE_UNKNOWN`, which would lead to the same outcome as the above `if self.is_unknown()` case.

OverloadUT replied:
The fan and light isy devices don't implement their own `state` property, so this base-class code would apply to them. I could move the logic in to those last types, but I like that having it in the base class also serves as a signal to future implementers that the UNKNOWN state needs to be handled if you override it.

This actually doesn't end up going to the `Entity` class! I had to test it a bunch because `super()`'s behavior is new to me, but `super` will actually call the implementation in the _other_ base class of the child object.

To use Fan as an example, it is defined as `ISYFanDevice(isy.ISYDevice, FanEntity)`. So `state` will first get handled by `ISYDevice.state`, and when it calls `super()` it kicks it on over to `FanEntity.state` which handles the fan-specific version of state.

OverloadUT commented:
@MartinHjelmare when you are done reviewing this PR, I'd love if you could toss the same message on to #10664, which this PR is branched off of (so all of the code in that PR is in the code you've been reviewing here)

MartinHjelmare commented:
If that's true what's the point of #10664?

OverloadUT commented:
Originally, #10664 was meant to be a small PR adding an important feature that things like this PR and others would need. But it took quite a while to get the new PyISY version all done and published, so in the meantime I worked on this PR as well.

I was trying to stick to the "smaller PRs" request, but this one depended on that other functionality so there was no choice but to branch from it. In a perfect world that other one would have been merged weeks ago :)

I suppose since you've been reviewing all of the code here, I could just close that other PR and roll both new features in to this one. I'll go with whatever is preferable on that front.

MartinHjelmare commented:
Yes, please close the other PR.

OverloadUT commented:
Done. Let me know when we're looking good here and I'll get the documentation written up! I wanted to wait until the implementation was locked-in :)

OverloadUT commented:
Re-adding WIP tag. I have discovered some issues with new devices I purchased for testing this change, and possibly with the latest ISY firmware 😭

OverloadUT committed:
Restrict negative-node and heartbeat support to known compatible types

Not all Insteon sensors use the same subnode IDs for the same things, so we need to use different logic depending on device type. Negative node and heartbeat support is now only used for leak sensors and open/close sensors.

OverloadUT commented:
Okay, should be ready for a final review again. I bought a motion sensor to validate these changes, and it turns out they work totally differently. I added a simple check to only apply the new logic in this PR to leak and door sensors.

I also improved initial state support so that doors with an unknown state on bootup will report as unknown until the sensor reports for the first time.

MartinHjelmare commented on homeassistant/components/binary_sensor/isy994.py:
```diff
@@ -82,6 +91,26 @@ def setup_platform(hass, config: ConfigType,
+
+    split_type = device_type.split('.')
+    for device_class, ids in ISY_DEVICE_TYPES.items():
+        if split_type[0] + '.' + split_type[1] in ids:
```
Use new style string formatting.
```py
'{}.{}'.format(split_type[0], split_type[1])
```

OverloadUT replied:
👍

OverloadUT committed:
Use new style string formatting

MartinHjelmare reviewed (approved):
Looks good!

OverloadUT commented:
Excellent; I'll write up the documentation now!

MartinHjelmare commented:
Add a link to a docs PR and I think we can merge this.

OverloadUT committed:
Add binary sensor detection for pre-5.x firmware

Meant to do this originally; writing documentation revealed that this requirement was missed!

OverloadUT commented:
Sorry for the one additional commit! In writing the documentation I realized that I never added the small bit of logic necessary for pre-5.x firmware to work, which was always my intention.

OverloadUT commented:
Description updated with a link to the documentation PR, and also added event support description here, which became part of this PR.
